packages/components/src/Tabs2/types.ts

```diff
@@ -0,0 +1,103 @@
+/*
+
+ MIT License
+
+ Copyright (c) 2021 Looker Data Sciences, Inc.
+
+ Permission is hereby granted, free of charge, to any person obtaining a copy
+ of this software and associated documentation files (the "Software"), to deal
+ in the Software without restriction, including without limitation the rights
+ to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ copies of the Software, and to permit persons to whom the Software is
+ furnished to do so, subject to the following conditions:
+
+ The above copyright notice and this permission notice shall be included in all
+ copies or substantial portions of the Software.
+
+ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ SOFTWARE.
+
+ */
+
+import { ReactElement } from 'react'
+import {
+  CompatibleHTMLProps,
+  FontSizeProps,
+  LayoutProps,
+  PaddingProps,
+  TypographyProps,
+} from '@looker/design-tokens'
+
+export type Tab2Props = Omit<CompatibleHTMLProps<HTMLButtonElement>, 'type'> &
+  LayoutProps &
+  PaddingProps &
+  TypographyProps & {
+    /**
+     * displays as the `Tab`'s value
+     */
+    label: string
+    /**
+     * callback to manage when `Tab` is clicked
+     */
+    onSelect?: () => void
+    /**
+     * specific `Tab` that is selected.
+     */
+    selected?: boolean
+  }
+
+export type TabList2Props = PaddingProps &
+  FontSizeProps & {
+    /**
+     * element that will be displayed as the `Tab` value
+     */
+    children: JSX.Element[]
+    className?: string
+    /**
+     * spread the collection of `Tab` on its full width.
+     */
+    distribute?: boolean
+  }
+
+type TabStackMember = Tab2Props & {
+  id: string
+}
+
+export type TabStack = TabStackMember[]
+
+type Controlled = {
+  /**
+   * Which tab to show on load
+   */
+  defaultTabId?: string
+  /**
+   * Controlled: which tab to show now
+   */
+  tabId?: string
+  /**
+   * Callback called when tabId changes
+   */
+  onTabChange?: (tabId: string) => void
+}
+
+type Uncontrolled = {
+  defaultTabId?: never
+  onTabChange?: never
+  tabId?: string
+}
+
+export type Tabs2Props = (Controlled | Uncontrolled) & {
```

Comment:
It looks like the intent here is to prevent users from using a `tabId` without also passing an `onTabChange` prop. That doesn't seem to work as expected! I can totally set `<Tabs2 tabId={currentTabId}>` or `<Tabs2 onTabChange={() => null}>` without typescript complaining. 

I took a bit to play around with the right type definitions, and I think I came up with a structure that requires you to do *both*  `tabId` and `onTabChange` OR do `defaultTabId`, but not any other combination. I think it should be structured something like this:

```suggestion
type Controlled = {
  /**
   * Controlled: which tab to show now
   */
  tabId: string
  /**
   * Callback called when tabId changes
   */
  onTabChange: (tabId: string) => void
  defaultTabId?: never
}

type Uncontrolled = {
  /**
   * Which tab to show on load
   */
  defaultTabId: string
  tabId?: never
  onTabChange?: never
}

type Default = {
  defaultTabId?: never
  tabId?: never
  onTabChange?: never
}

export type Tabs2Props = (Controlled | Uncontrolled | Default) & {
```

CC'ing @lukelooker to double check my logic there

Reply:
Oh man I always get so confused on the props that should and should not be a part of controlled or uncontrolled. 
You are right what I had didn't quite work , but I think both `Uncontrolled` and `Controlled` should have the possibility of passing `defaultTabId` and to be honest the `Default` option gets me even more confused. Seems like what you tried and what I tried is still not quite the right solution, I'm getting a TS error with ether variaton.

`
  Type '{ children: Element[]; defaultTabId?: string | undefined; tabId?: string | undefined; onTabChange?: ((tabId: string) => void) | undefined; distributed?: boolean | undefined; } | { ...; }' is not assignable to type 'IntrinsicAttributes & Tabs2Props'.
    Type '{ children: Element[]; defaultTabId?: string | undefined; tabId?: string | undefined; onTabChange?: ((tabId: string) => void) | undefined; distributed?: boolean | undefined; }' is not assignable to type 'IntrinsicAttributes & Tabs2Props'.
      Type '{ children: Element[]; defaultTabId?: string | undefined; tabId?: string | undefined; onTabChange?: ((tabId: string) => void) | undefined; distributed?: boolean | undefined; }' is not assignable to type 'Uncontrolled'.
      Types of property 'defaultTabId' are incompatible.
      Type 'string | undefined' is not assignable to type 'undefined'.
      Type 'string' is not assignable to type 'undefined'.
  `

Reply:
Ugh yeah this is really hard to type! Where is that error coming from?

About `defaultTabId` vs `tabId`, I'm pretty sure the best practice would be that you can't use them both simultaneously. My thinking is that if the values conflict, which one should the component implement? 

I looked up some articles about doing `value` vs `defaultValue` in forms, and they seem to confirm that the pattern should be mutually exclusive

https://scriptverse.academy/tutorials/reactjs-defaultvalue-value.html

<img width="681" alt="Screen Shot 2021-08-17 at 4 32 16 PM" src="https://user-images.githubusercontent.com/238293/129813962-f57799c5-5ba3-4adf-8c80-2b07a43f7010.png">

Reply:
>  to be honest the Default option gets me even more confused.

Yeah I totally understand that! I added it to account for the use case where they don't pass either `tabId` or `defaultTabId`. The thinking is that there are actually three possible scenarios here:

1. User can pass BOTH `tabId` and `onTabChange` (to require both you have to remove the optional `?` setting from those props)
2. User can pass ONLY `defaultTabId`
3. The user passes none of the above options. <-- that's the case that I added the `Default` option to describe. I agree that it seems like it might not be necessary, but it seemed to help in my local tests 🤷‍♂️